agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+        "        if (is_array($root_package) && array_key_exists('name', $root_package) && $packageName == $root_package['name']) {"
+        "          continue;"
+        "        }"
+        "        if (isset($packageData['pretty_version'])) {"
+        "          $packages[$packageName] = ltrim($packageData['pretty_version'], 'v');"
+        "        }"
+        "      }"
+        "    }"
+        "    return $packages;"
+        "  } catch (Exception $e) {"
+        "    return NULL;"
+        "  }"
+        "})();";
+  // clang-format on
+
+  if (NR_SUCCESS != nr_execute_handle_autoload_composer_init(vendor_path)) {
+    nrl_debug(NRL_INSTRUMENT,
+              "%s - unable to initialize Composer runtime API - package info "
+              "unavailable",
+              __func__);
+    return;
+  }
+
+  nrl_verbosedebug(NRL_INSTRUMENT, "%s - Composer runtime API available",
+                   __func__);
+
+  result
+      = zend_eval_string(getallrawdata, &retval, "composer_getallrawdata.php");
+  if (SUCCESS != result) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - composer_getallrawdata.php failed",
+                     __func__);
+    return;
+  }
+  if (IS_ARRAY == Z_TYPE(retval)) {
+    zend_string* package_name = NULL;
+    zval* package_version = NULL;
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL(retval), package_name,
+                                  package_version) {
+      if (NULL == package_name || NULL == package_version) {
+        continue;
+      }
+      if (nr_php_is_zval_non_empty_string(package_version)) {
+        nrl_verbosedebug(NRL_INSTRUMENT, "package %s, version %s",
+                         NRSAFESTR(ZSTR_VAL(package_name)),
+                         NRSAFESTR(Z_STRVAL_P(package_version)));
+        nr_txn_add_php_package_from_source(
+            NRPRG(txn), NRSAFESTR(ZSTR_VAL(package_name)),
+            NRSAFESTR(Z_STRVAL_P(package_version)),
+            NR_PHP_PACKAGE_SOURCE_COMPOSER);
+      }
+    }
+    ZEND_HASH_FOREACH_END();
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - installed packages is not an array",
+                     __func__);
+  }
+  zval_dtor(&retval);
+}
+
+static char* nr_execute_handle_autoload_composer_get_vendor_path(
+    const char* filename) {
+  char* vendor_path = NULL;  // result of dirname(filename)
+  char* cp = NULL;
+
+  // vendor_path = dirname(filename):
+  // 1. copy filename to vendor_path
+  vendor_path = nr_strdup(filename);
+  // 2. // find last occurence of '/' in vendor_path
+  cp = nr_strrchr(vendor_path, '/');
+  // 3. replace '/' with '\0' to get the directory path
+  *cp = '\0';
+
+  return vendor_path;
+}
+
+static bool nr_execute_handle_autoload_composer_file_exists(
+    const char* vendor_path,
+    const char* filename) {
+  char* composer_magic_file = NULL;  // vendor_path + filename
+  bool file_exists = false;
+
+  composer_magic_file = nr_formatf("%s/%s", vendor_path, filename);
```

Comment:
Is there a check for filename != NULL? otherwise, this will generate an error.

Reply:
`filename` is guaranteed to be not `NULL` here by [this](https://github.com/newrelic/newrelic-php-agent/blob/feat/vm/user-composer-for-package-info/agent/php_execute.c#L1026-L1032). However, I'll add `NULL` checks for `filename`.

Reply:
Added in cb83165d367509f3bb3439ad14b0aa61a21225c3.